src/OrbitGl/PagefaultTrack.cpp

```diff
@@ -0,0 +1,137 @@
+  if (!minor_pagefault_track_->IsEmpty()) result.push_back(minor_pagefault_track_.get());
+  return result;
+}
+
+void PagefaultTrack::Draw(Batcher& batcher, TextRenderer& text_renderer,
+                          uint64_t current_mouse_time_ns, PickingMode picking_mode,
+                          float z_offset) {
+  float track_height = GetHeight();
+  float track_width = viewport_->GetVisibleWorldWidth();
+
+  SetPos(viewport_->GetWorldTopLeft()[0], pos_[1]);
+  SetSize(track_width, track_height);
+
+  UpdatePositionOfSubtracks();
+
+  Track::Draw(batcher, text_renderer, current_mouse_time_ns, picking_mode, z_offset);
+
+  if (collapse_toggle_->IsCollapsed()) return;
+
+  if (!major_pagefault_track_->IsEmpty()) {
+    major_pagefault_track_->SetSize(viewport_->GetVisibleWorldWidth(),
+                                    major_pagefault_track_->GetHeight());
+    major_pagefault_track_->Draw(batcher, text_renderer, current_mouse_time_ns, picking_mode,
+                                 z_offset);
+  }
+
+  if (!minor_pagefault_track_->IsEmpty()) {
+    minor_pagefault_track_->SetSize(viewport_->GetVisibleWorldWidth(),
+                                    minor_pagefault_track_->GetHeight());
+    minor_pagefault_track_->Draw(batcher, text_renderer, current_mouse_time_ns, picking_mode,
+                                 z_offset);
+  }
+}
+
+void PagefaultTrack::UpdatePrimitives(Batcher* batcher, uint64_t min_tick, uint64_t max_tick,
+                                      PickingMode picking_mode, float z_offset) {
+  UpdatePositionOfSubtracks();
+
+  if (!major_pagefault_track_->IsEmpty()) {
+    major_pagefault_track_->UpdatePrimitives(batcher, min_tick, max_tick, picking_mode, z_offset);
+  }
+
+  if (collapse_toggle_->IsCollapsed()) return;
+
+  if (!minor_pagefault_track_->IsEmpty()) {
+    minor_pagefault_track_->UpdatePrimitives(batcher, min_tick, max_tick, picking_mode, z_offset);
+  }
+}
+
+void PagefaultTrack::UpdatePositionOfSubtracks() {
+  if (collapse_toggle_->IsCollapsed()) {
+    major_pagefault_track_->SetPos(pos_[0], pos_[1]);
+    return;
+  }
+
+  float current_y = pos_[1] - layout_->GetTrackTabHeight();
+  if (!major_pagefault_track_->IsEmpty()) {
+    current_y -= layout_->GetSpaceBetweenSubtracks();
+  }
+  major_pagefault_track_->SetPos(pos_[0], current_y);
+
+  if (!minor_pagefault_track_->IsEmpty()) {
+    current_y -= (layout_->GetSpaceBetweenSubtracks() + major_pagefault_track_->GetHeight());
+  }
+  minor_pagefault_track_->SetPos(pos_[0], current_y);
+}
+
+void PagefaultTrack::OnTimer(const orbit_client_protos::TimerInfo& timer_info) {
+  constexpr uint32_t kDepth = 0;
+  std::shared_ptr<TimerChain> timer_chain = timers_[kDepth];
+  if (timer_chain == nullptr) {
+    timer_chain = std::make_shared<TimerChain>();
+    timers_[kDepth] = timer_chain;
+  }
+
+  timer_chain->emplace_back(timer_info);
+}
+
+std::vector<std::shared_ptr<TimerChain>> PagefaultTrack::GetAllChains() const {
```

Comment:
Why we need to overwrite and reimplement this method?

Reply:
This is a pure virtual method in `Track`, I think we have to overwrite it.